### PR TITLE
P32

### PR DESCRIPTION
**Pull request by lucasperin**

### CODE REVIEW

File: p32/src/bank.rs [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    pub name: String,
+    pub users: Vec<User>,
+    pub credit_interst: u64,
+    pub debit_interst: u64,
+}
+
+#[derive(Debug, Default, Eq, PartialEq)]
+pub struct BankBalance {
+    pub liabilities: i64,
+    pub assets: i64,
```
Comment (newpavlov) [inline p32/src/bank.rs:22]: Assets and liabilities are always positive, so it's better to use `u64`. You can use `unsigned_abs` for converting `i64` to `u64`.

File: p32/src/bank.rs [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+                .sum(),
+        }
+    }
+
+    pub fn transfer_funds(
+        &mut self,
+        amount: u64,
+        origin: &str,
+        destination: &str,
+    ) -> Result<(), Error> {
```
Comment (newpavlov) [inline p32/src/bank.rs:52]: You use `std::fmt::Error` type. It has nothing to do with this method. You need to define your own error type. I recommend to use an `enum` with different failure cases..

File: p32/src/bank.rs [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        let p1 = self.user_index_by_name(origin);
+        let p2 = self.user_index_by_name(destination);
+        match (p1, p2) {
+            (Some(p1), Some(p2)) => {
+                if self.users[p1].balance >= amount as i64 {
```
Comment (newpavlov) [inline p32/src/bank.rs:57]: You need to check against user's credit limit, not against zero. In other words, users can have negative balances as long as it's not bellow their limit.

File: p32/src/bank.rs [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+                    self.users[p1].balance -= amount as i64;
+                    self.users[p2].balance -= amount as i64;
```
Comment (newpavlov) [inline p32/src/bank.rs:59]: You remove funds from both accounts. This code also does not account for potential overflow and underflow conditions. For example, if `amount` is bigger than 2^63 the `as` cast will produce a huge negative number.

File: p32/src/bank.rs [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+                    Err(Error)
+                }
+            }
+            _ => Err(Error),
+        }
+    }
+
+    pub fn accrue_interes(&mut self) {
+        for user in self.users.iter_mut().filter(|user| user.balance > 0) {
+            user.balance = ((user.balance * self.debit_interst as i64) as f32 / 100f32) as i64;
```
Comment (newpavlov) [inline p32/src/bank.rs:71]: Using floats for financial computations is a big no-no. You need to implement it using strictly integer arithmetic. Also, the formula itself is not correct. Remember that interests are given in basis points.

File: p32/src/bank.rs [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+    }
+
+    pub fn merge_bank(&mut self, bank: Bank) {
+        for user in bank.users.iter() {
+            match self.user_index_by_name(user.name.as_str()) {
+                Some(position) => {
+                    self.users[position].balance += user.balance;
+                }
+                None => {
+                    self.users.push(user.clone());
```
Comment (newpavlov) [inline p32/src/bank.rs:85]: You can remove clone by iterating over owned users, i.e. by using `for user in bank.user { .. }`.

File: p32/src/bank.rs (hunk 3 above)
Comment (lucasperin) [reply p32/src/bank.rs:57]: I did not understand this part of the exercise exactly.

**Pull request merged**